text/0000-macros-1.1.md

```diff
@@ -0,0 +1,564 @@
+only be converted to and from a `String`. Eventually `TokenStream` type will
+more closely resemble token streams [in the compiler
+itself][compiler-tokenstream], and more fine-grained manipulations will be
+available as well.
+
+Additionally, the `Context` structure will initially be completely devoid
+of functionality, but in the future it will be the entry point for [many other
+features][macro20] one would expect in macros 2.0
+
+[compiler-tokenstream]: https://github.com/rust-lang/rust/blob/master/src/libsyntax/tokenstream.rs#L323-L338
+[macro20]: http://ncameron.org/blog/libmacro/
+
+### Defining a macro
+
+A new crate type will be added to the compiler, `rustc-macro` (described below),
+indicating a crate that's compiled as a procedural macro. There will not be a
+"registrar" function in this crate type (like there is today), but rather a
+number of functions which act as token stream transformers to implement macro
+functionality.
+
+A macro crate might look like:
+
+```rust
+#![crate_type = "rustc-macro"]
+#![crate_name = "double"]
+
+extern crate rustc_macro;
+
+use rustc_macro::{Context, TokenStream};
+
+#[rustc_macro_derive(Double)]
+pub fn double(cx: &mut Context, input: TokenStream) -> TokenStream {
+    let source = input.to_source(cx);
+
+    // Parse `source` for struct/enum declaration, and then build up some new
+    // source code representing representing a number of items in the
+    // implementation of the `Double` trait for the struct/enum in question.
+    let source = derive_double(cx, source);
+
+    // Parse this back to a token stream and return it
+    TokenStream::from_source(cx, &source).unwrap()
+}
+```
+
+This new `rustc_macro_derive` attribute will be allowed inside of a
+`rustc-macro` crate but disallowed in other crate types. It defines a new
+`#[derive]` mode which can be used in a crate. The input here is the entire
+struct that `#[derive]` was attached to, attributes and all. The output is
+**expected to include the `struct`/`enum` itself** as well as any number of
+items to be contextually "placed next to" the initial declaration.
+
+Again, though, there is no hygiene, it's as if the source was simply
+copy/pasted. All span information for the `TokenStream` structures returned by
+`from_source` will point to the original `#[derive]` annotation. This means
+that error messages related to struct definitions will get *worse* if they have
+a custom derive attribute placed on them, because the entire struct's span will
+get folded into the `#[derive]` annotation. Eventually, though, more span
+information will be stable on the `TokenStream` type, so this is just a
+temporary limitation.
+
+The `rustc_macro_derive` attribute requires the signature (similar to [macros
+2.0][mac20sig]):
+
+[mac20sig]: http://ncameron.org/blog/libmacro/#tokenisingandquasiquoting
+
+```rust
+fn(&mut Context, TokenStream) -> TokenStream
+```
+
+If a macro cannot process the input token stream, it is expected to panic for
```

Comment:
I believe the intention here is that this is the signature we'll want in "macros 2.0" as well, and eventually diagnostics should be signaled through `rustc_macro`, not panics. That is, any `Result` type returned here wouldn't have enough information for diagnostics anyway.